flink-ml-core/src/main/java/org/apache/flink/ml/linalg/typeinfo/DenseMatrixSerializer.java

```diff
@@ -0,0 +1,138 @@
+
+    private static final double[] EMPTY = new double[0];
+
+    private static final DenseMatrixSerializer INSTANCE = new DenseMatrixSerializer();
+
+    @Override
+    public boolean isImmutableType() {
+        return false;
+    }
+
+    @Override
+    public DenseMatrix createInstance() {
+        return new DenseMatrix(0, 0, EMPTY);
+    }
+
+    @Override
+    public DenseMatrix copy(DenseMatrix from) {
+        return new DenseMatrix(
+                from.numRows, from.numCols, Arrays.copyOf(from.values, from.values.length));
+    }
+
+    @Override
+    public DenseMatrix copy(DenseMatrix from, DenseMatrix reuse) {
+        if (from.values.length == reuse.values.length) {
+            System.arraycopy(from.values, 0, reuse.values, 0, from.values.length);
+            return reuse;
+        }
+        return copy(from);
+    }
+
+    @Override
+    public int getLength() {
+        return -1;
+    }
+
+    @Override
+    public void serialize(DenseMatrix matrix, DataOutputView target) throws IOException {
+        if (matrix == null) {
+            throw new IllegalArgumentException("The matrix must not be null.");
+        }
+
+        final int len = matrix.values.length;
+        target.writeInt(matrix.numRows);
+        target.writeInt(matrix.numCols);
+        for (int i = 0; i < len; i++) {
+            target.writeDouble(matrix.values[i]);
+        }
+    }
+
+    @Override
+    public DenseMatrix deserialize(DataInputView source) throws IOException {
+        int m = source.readInt();
+        int n = source.readInt();
+        double[] values = new double[m * n];
+        for (int i = 0; i < m * n; i++) {
+            values[i] = source.readDouble();
+        }
+        return new DenseMatrix(m, n, values);
+    }
+
+    private static void readDoubleArray(double[] dst, DataInputView source, int len)
+            throws IOException {
+        for (int i = 0; i < len; i++) {
+            dst[i] = source.readDouble();
+        }
+    }
+
+    @Override
+    public DenseMatrix deserialize(DenseMatrix reuse, DataInputView source) throws IOException {
+        int m = source.readInt();
+        int n = source.readInt();
+
+        double[] values = new double[m * n];
+        readDoubleArray(values, source, m * n);
+        return new DenseMatrix(m, n, values);
+    }
+
+    @Override
+    public void copy(DataInputView source, DataOutputView target) throws IOException {
+        int m = source.readInt();
+        target.writeInt(m);
+        int n = source.readInt();
+        target.writeInt(n);
+
+        target.write(source, m * n * 8);
+    }
+
+    // ------------------------------------------------------------------------
+
+    @Override
+    public TypeSerializerSnapshot<DenseMatrix> snapshotConfiguration() {
+        return new DenseMatrixSerializerSnapshot();
+    }
+
+    /** Serializer configuration snapshot for compatibility and format evolution. */
+    @SuppressWarnings("WeakerAccess")
```

Comment:
what is the functionality of this method?

Reply:
@lindong28 I write this method as VectorSerializer. Can you explain the functionality?

Reply:
@weibozhao I think @zhipeng93 is asking why we need "WeakerAccess".

@zhipeng93 According to [1], it seems that `@SuppressWarnings("WeakerAccess")` is related to the `can be private` warning. This line of code is copied from existing Flink code (e.g. `BigDecSerializerSnapshot`). IMO it is OK to just follow the existing practice unless we have clear reason why it is no longer needed.

[1] https://stackoverflow.com/questions/38628679/what-value-for-suppresswarnings-do-i-use-to-suppress-the-can-be-private-warni